examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py

```diff
@@ -82,109 +71,60 @@
 """
 
 
+def use_inference_server(cfg, model, trainer):
+    if not HAVE_MEGATRON_CORE:
+        raise ValueError('Megatron-core needs to be installed to use this feature!')
+
+    from nemo.collections.nlp.modules.common.megatron_web_server import get_chatbot_demo, get_demo
+
+    trainer.test(model, dataloaders=None)
+
+    if parallel_state.is_pipeline_first_stage() and parallel_state.get_tensor_model_parallel_rank() == 0:
+        if cfg.web_server:
+            if cfg.chat:
+                defaults = {
+                    'user': cfg.chatbot_config.user,
+                    'assistant': cfg.chatbot_config.assistant,
+                    'system': cfg.chatbot_config.system,
+                }
+                web_ui = partial(
+                    get_chatbot_demo,
+                    defaults=defaults,
+                    value=cfg.chatbot_config.value,
+                    attributes=cfg.chatbot_config.attributes,
+                )
+            else:
+                web_ui = get_demo
+            loop = asyncio.new_event_loop()
+            thread = threading.Thread(
+                target=web_ui, daemon=True, args=(cfg.share, cfg.username, cfg.password, cfg.port, cfg.web_port, loop),
+            )
+            thread.start()
+        server = MegatronServer(model.cuda())
+        server.run("0.0.0.0", port=cfg.port)
+
+    while True:
+        choice = torch.cuda.LongTensor(1)
+        torch.distributed.broadcast(choice, 0)
+        if choice[0].item() == 0:
+            generate(model.cuda())
+
+
 @hydra_runner(config_path="conf", config_name="megatron_gpt_peft_eval_config")
 def main(cfg) -> None:
     logging.info("\n\n************** Experiment configuration ***********")
     logging.info(f"\n{OmegaConf.to_yaml(cfg)}")
-    assert cfg.model.restore_from_path is not None
-    megatron_amp_o2 = cfg.model.get("megatron_amp_O2", False)
-    with_distributed_adam = False
-
-    plugins = []
-    strategy = NLPDDPStrategy(
-        no_ddp_communication_hook=True,  # we don't use DDP for async grad allreduce
-        gradient_as_bucket_view=cfg.model.gradient_as_bucket_view,
-        find_unused_parameters=False,
-    )
-    if cfg.trainer.precision in [16, '16', 'bf16', '16-mixed', 'bf16-mixed']:
-        scaler = None
-        if cfg.trainer.precision in [16, '16', '16-mixed']:
-            scaler = GradScaler(
-                init_scale=cfg.model.get("native_amp_init_scale", 2 ** 32),
-                growth_interval=cfg.model.get("native_amp_growth_interval", 1000),
-                hysteresis=cfg.model.get("hysteresis", 2),
-                enabled=False
-                if cfg.model.pipeline_model_parallel_size > 1
```

Comment:
Just fyi, this `enabled` flag is not set in the TrainerBuilder. I'm not sure if it's essential, maybe @ericharper knows. When I've been using the builder, I've tried to make sure existing code matches line for line and using subclasses when necessary to override some functions. Maybe we can also compare what is in the Builder with other examples to see if that needs to be changed.

Reply:
Same question/comment for tuning example script

Reply:
Thanks for catching this! PP is not tested right now but will probably be needed in the future. I'll create a subclass for this to match the old script